### PR TITLE
API GetTransactionProof, GetReceiptProof

### PR DESCRIPTION
**Pull request by wgr523**

### CODE REVIEW

**Review by GalaxySciTech — COMMENTED** [review]
it need a test script

**Review by liam-lai — APPROVED** [review]

**Comment by liam-lai** [comment]
Yes, please write a test for proofPairList object and deriveTrie function at least

**Review by liam-lai — APPROVED** [review]

**Comment by wanwiset25** [comment]
do we need to put something in web3ext.go for external interface? 
something like the below
`		new web3._extend.Method({
			name: 'getV2Block',
			call: function (args) {
				return (web3._extend.utils.isString(args[0]) && args[0].indexOf('0x') === 0) ? "XDPoS_getV2BlockByHash" : "XDPoS_getV2BlockByNumber";
			},
			params: 1,
			inputFormatter: [web3._extend.formatters.inputBlockNumberFormatter]`

File: internal/ethapi/receipt_proof_test.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+			if err != nil {
+				return nil, err
+			}
+			return b, nil
+		}
+	}
+	return nil, fmt.Errorf("key not found")
+}
+
+func TestReceiptProof(t *testing.T) {
```
Comment (GalaxySciTech) [inline internal/ethapi/receipt_proof_test.go]: i think it is necessary to call  GetReceiptProof function in test script 

we need test the function

**Comment by GalaxySciTech** [comment]
it seems Receipt need to add a parameter `contractAddress`  in here https://github.com/XinFinOrg/XDC-Subnet/blob/36431045571085086727850d5cab0b5d06b02ed8/core/types/receipt.go#L99

File: internal/ethapi/receipt_proof_test.go (hunk 1 above)
Comment (wgr523) [reply internal/ethapi/receipt_proof_test.go]: I don't know how to write test for calling function like GetReceiptProof or other GetXXX. In the codebase I cannot find any test like this.

Comment (GalaxySciTech) [reply internal/ethapi/receipt_proof_test.go]: if use that way how do we ensure the getReceiptProof will no error when we changed some code

**Comment by wgr523** [comment]
> it seems Receipt need to add a parameter `contractAddress` in here
> 
> https://github.com/XinFinOrg/XDC-Subnet/blob/36431045571085086727850d5cab0b5d06b02ed8/core/types/receipt.go#L99

This encoding is written by ethereum yellow paper and I think we should not change it. If we change it, then our code is quite different from ethereum and users may not like it.

The code to compute contractAddress is quite simple and we can compute it whenever we want to use it: in `core/state_processor.go`
```
	// if the transaction created a contract, store the creation address in the receipt.
	if msg.To() == nil {
		receipt.ContractAddress = crypto.CreateAddress(vmenv.Context.Origin, tx.Nonce())
	}
```
(vmenv.Context.Origin is just tx.Sender())

**Comment by GalaxySciTech** [comment]
> > it seems Receipt need to add a parameter `contractAddress` in here
> > https://github.com/XinFinOrg/XDC-Subnet/blob/36431045571085086727850d5cab0b5d06b02ed8/core/types/receipt.go#L99
> 
> This encoding is written by ethereum yellow paper and I think we should not change it. If we change it, then our code is quite different from ethereum and users may not like it.
> 
> The code to compute contractAddress is quite simple and we can compute it whenever we want to use it: in `core/state_processor.go`
> 
> ```
> 	// if the transaction created a contract, store the creation address in the receipt.
> 	if msg.To() == nil {
> 		receipt.ContractAddress = crypto.CreateAddress(vmenv.Context.Origin, tx.Nonce())
> 	}
> ```
> 
> (vmenv.Context.Origin is just tx.Sender())

i can only get the receiptrlp 

the struct only have status bloom log gas 

i can not get other parameters

File: internal/ethapi/receipt_proof_test.go (hunk 1 above)
Comment (wgr523) [reply internal/ethapi/receipt_proof_test.go]: actually I don't know. I can only say that in our codebase (and the ethereum codebase xinfin forked) there is no test for api functions. So it is hard to write test for this api function `GetReceiptProof `

**Comment by GalaxySciTech** [comment]
i think if any possible we can implement the `eth_getProof` function? 
this is the reference https://eips.ethereum.org/EIPS/eip-1186

**Comment by wgr523** [comment]
> i think if any possible we can implement the `eth_getProof` function? this is the reference https://eips.ethereum.org/EIPS/eip-1186

I find the implementation and can do it. But may I ask when do we need eth_getProof? I thought we only need transaction proof and receipt proof.

**Comment by GalaxySciTech** [comment]
> > i think if any possible we can implement the `eth_getProof` function? this is the reference https://eips.ethereum.org/EIPS/eip-1186
> 
> I find the implementation and can do it. But may I ask when do we need eth_getProof? I thought we only need transaction proof and receipt proof.

just forget it , this one just is a another way that deal the cross chain system

File: internal/ethapi/api.go [hunk 2]
```diff
@@ -534,6 +534,49 @@ func (s *PublicBlockChainAPI) GetBalance(ctx context.Context, address common.Add
...
+	}
+	if len(receipts) <= int(index) {
+		return nil, nil
+	}
+	receipt_tr := deriveTrie(receipts)
+	var receipt_proof proofPairList
+	if err := receipt_tr.Prove(keybuf.Bytes(), 0, &receipt_proof); err != nil {
+		return nil, err
+	}
+	fields := map[string]interface{}{
```
Comment (GalaxySciTech) [inline internal/ethapi/api.go:567]: add stateProof

Comment (wgr523) [reply internal/ethapi/api.go:567]: hi, I thought you said just forget it... What is stateProof's input that you need? a transaction is not in the stateProof, only account balance, nonce etc are in stateProof

Comment (GalaxySciTech) [reply internal/ethapi/api.go:567]: This is only added for future use. If you need to use this interface directly, you don’t need to write it again.

Comment (wgr523) [reply internal/ethapi/api.go:567]: Got it. Can we do it in another PR?

Comment (GalaxySciTech) [reply internal/ethapi/api.go:567]: sure

**Pull request merged**